### ISSUE TITLE
`<<loop>>` while generating doc of svg-tree with GHC 8.2.2

### ISSUE THREAD

**Issue by Twinside** [haskell/haddock#710]
Using head of [svg-tree](https://github.com/Twinside/svg-tree) (equivalent to the one on [hackage](https://hackage.haskell.org/package/svg-tree) ) we get the following result:

```
[11:28:10] ~/Documents/Coding/svg-tree> stack haddock . --resolver=nightly-2017-11-25
... (lots of generation)
Running Haddock on library for svg-tree-0.6.2..
links will be generated to these packages: fail-4.9.0.0
Haddock coverage:
 100% ( 16 / 16) in 'Graphics.Svg.CssTypes'
   0% (  0 /  2) in 'Graphics.Svg.NamedColors'
  Missing documentation for:
    Module header
    svgNamedColors (src\Graphics\Svg\NamedColors.hs:8)
  97% (131 /135) in 'Graphics.Svg.Types'
  Missing documentation for:
    MeshGradientType (src\Graphics\Svg\Types.hs:228)
    HasMeshGradient (src\Graphics\Svg\Types.hs:844)
    HasMeshGradientRow (src\Graphics\Svg\Types.hs:820)
    HasMeshGradientPatch (src\Graphics\Svg\Types.hs:808)
   0% (  0 / 12) in 'Graphics.Svg.PathParser'
  Missing documentation for:
    Module header
    transformParser (src\Graphics\Svg\PathParser.hs:162)
    command (src\Graphics\Svg\PathParser.hs:69)
    pathParser (src\Graphics\Svg\PathParser.hs:66)
    viewBoxParser (src\Graphics\Svg\PathParser.hs:48)
    pointData (src\Graphics\Svg\PathParser.hs:63)
    gradientCommand (src\Graphics\Svg\PathParser.hs:234)
    serializePoints (src\Graphics\Svg\PathParser.hs:110)
    serializeCommand (src\Graphics\Svg\PathParser.hs:132)
    serializeGradientCommand (src\Graphics\Svg\PathParser.hs:248)
    serializeCommands (src\Graphics\Svg\PathParser.hs:129)
    serializeViewBox (src\Graphics\Svg\PathParser.hs:54)
   0% (  0 /  6) in 'Graphics.Svg.ColorParser'
  Missing documentation for:
    Module header
    colorParser (src\Graphics\Svg\ColorParser.hs:56)
    colorSerializer (src\Graphics\Svg\ColorParser.hs:53)
    textureParser (src\Graphics\Svg\ColorParser.hs:105)
    textureSerializer (src\Graphics\Svg\ColorParser.hs:94)
    urlRef (src\Graphics\Svg\ColorParser.hs:99)
  20% (  2 / 10) in 'Graphics.Svg.CssParser'
  Missing documentation for:
    Module header
    complexNumber (src\Graphics\Svg\CssParser.hs:204)
    declaration (src\Graphics\Svg\CssParser.hs:171)
    ruleSet (src\Graphics\Svg\CssParser.hs:124)
    styleString (src\Graphics\Svg\CssParser.hs:131)
    dashArray (src\Graphics\Svg\CssParser.hs:198)
    numberList (src\Graphics\Svg\CssParser.hs:201)
    num (src\Graphics\Svg\CssParser.hs:61)
  20% (  1 /  5) in 'Graphics.Svg.XmlParser'
  Missing documentation for:
    Module header
    unparseDocument (src\Graphics\Svg\XmlParser.hs:1175)
    SvgAttributeLens (src\Graphics\Svg\XmlParser.hs:373)
    drawAttributesList (src\Graphics\Svg\XmlParser.hs:561)
 100% ( 13 / 13) in 'Graphics.Svg'
Warning: Graphics.Svg.CssTypes: could not find link destinations for:
    TextBuildable
haddock: internal error: <<loop>>
Completed 40 action(s).

--  While building package svg-tree-0.6.2 using:
      C:\sr\setup-exe-cache\x86_64-windows\Cabal-simple_Z6RU0evB_2.0.1.0_ghc-8.2.2.exe --builddir=.stack-work\dist\5c8418a7 haddock --html --hoogle --html-location=../$pkg-$version/ --haddock-option=--hyperlinked-source
    Process exited with code: ExitFailure 1
```

This ticket is related to Twinside/svg-tree#11 . While this is a bug, I'd be interested in any information to help me work-around it, thanks :)

**Comment by Twinside** [comment]
It looks like it may be related to #679

**Comment by PierreR** [comment]
As a note, in #679 there is no such output `<<loop>>`.

**Comment by harpocrates** [comment]
Here is a stripped down test case (unfortunately still depending on `lens`) of something that doesn't terminate. I was never able to reproduce the `<<loop>>` output.

```haskell
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE FlexibleInstances #-}
-- | This module define all the types used in the definition
-- of a svg scene.
--
-- Most of the types are lensified.
module Graphics.Svg.Types where

import Control.Lens.TH ( makeClassy )

-- | Define a SVG group, corresponding `<g>` tag.
data Group a = Group

-- | Lenses associated to the Group type.
makeClassy ''Group
```

**Comment by PierreR** [comment]
>  I was never able to reproduce the <<loop>> output.

This is #679 (which is related to `makeClassy`.

@harpocrates Thanks for the stripped down test case !

**Comment by alexbiehl** [comment]
This sounds like a TemplateHaskell issue. ghc-8.4 alpha has been released recently, could you try again with the provided haddock? There has been some work around handling TemplateHaskell there.

**Comment by harpocrates** [comment]
I've been able to distill down a dependency-free (apart from `template-haskell`) test case. In a nutshell, I swapped out `lens` for `micro-lens` and `micro-lens-th`, inlined the required definitions and hacked stuff out. Using this test case, I bisected and fell on 506f614402192bd7b6a9a608e925a01b373b2bdc. :/

**Comment by harpocrates** [comment]
I think I understand the problem: the nifty new `specialize` function from [here](https://github.com/haskell/haddock/commit/506f614402192bd7b6a9a608e925a01b373b2bdc#diff-8534729b33620e15585044ff43ef0e2cR36) makes some assumptions about the freshness of type variables. Those assumptions seem to hold for any code GHC would parse, but not for any code that TH can generate.

The `makeClassy` issue stems from the fact that `makeClassy` reuses the same type variables in the class and instance head. Here is a shorter test-case which also hangs. Notice how giving the instance  head a fresh type variable fixes the non-termination.

```haskell
{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -ddump-splices #-}

module Foo where

import Language.Haskell.TH

data Bar a = Bar

$(do
   a <- newName "a"
   
   let classN = mkName "Foo"
   let methodN = mkName "foo"

   methodTy <- [t| $(varT a) -> $(varT a) |]
   let cla = ClassD [] classN [PlainTV a] [] [SigD methodN methodTy]
 
   -- uncommenting the next line makes Haddock terminate
   -- a <- newName "a"

   instanceHead <- [t| $(conT classN) (Bar $(varT a)) |]
   idCall <- [e| id |]
   let ins = InstanceD Nothing [] instanceHead [FunD methodN [Clause [] (NormalB idCall) []]]
    
   pure [cla,ins])
```

**Comment by alexbiehl** [comment]
Fixed by #723
